### DISCUSSION TITLE
2d LINE and ARC into 3d extruded 3d model ?

### DISCUSSION THREAD

**Discussion by manpowre** [mozman/ezdxf#1038] [General]
I have been looking into ezdxf last few days, got to know LINE, ARC, LWPOLYLINE, etc... pretty well. made code to open a dxf file, recreate LINE and ARC's into a new file, even converting it to lwpolyline. My goal is to make this so the 2d drawing which comes from locked software into a 3d model as the output changes every day.

I thought just to ask the community if what I am trying to achieve is possible at all ?

From this LINE and ARC's only:
<img width="671" alt="Screenshot 2024-02-25 104914" src="https://github.com/mozman/ezdxf/assets/29288793/66f0e62c-7793-4c9d-a608-650ffe484434">

to this:
<img width="599" alt="Screenshot 2024-02-25 105047" src="https://github.com/mozman/ezdxf/assets/29288793/57069503-a372-42f0-be62-1b993af839db">

I have made a function that converts the line's into lwpolyline with z as the height factor (either start or end + height). I slightly looked into arc, but I figured if I cant extrude the line's into a 3d face, then I wont dig further. When I open up the new dxf, the lwpolylines is not extruded in the z axis.

`
        for entity in msp:
        if entity.dxftype() == "LINE":
            start = entity.dxf.start
            end = entity.dxf.end
            start_3d = (start[0], start[1], start[2] + 1)
            end_3d = (end[0], end[1], end[2] + 1)

            polyline= new_msp.add_lwpolyline([start_3d, end_3d], dxfattribs={
                "layer": entity.dxf.layer,
                "color": entity.dxf.color,
                "linetype": entity.dxf.linetype,
                "lineweight": entity.dxf.lineweight,                
            })
`

Well actually closing the top and bottom is next step.

What do you all think, possible ?

**Comment by manpowre** [comment]
It dawned on me upon writing the above. make faces out of 2 polylines with vertice1 and vertice2. good to have someone to talk to. 
<img width="710" alt="Screenshot 2024-02-25 121253" src="https://github.com/mozman/ezdxf/assets/29288793/66b923b6-370f-476d-b746-c0c2e25b2e9a">

def convert_to_polyline(input_file, output_file, thickness):
    doc = ezdxf.readfile(input_file)
    new_doc = ezdxf.new()
    msp = doc.modelspace()
    new_msp = new_doc.modelspace()
    mesh = new_msp.add_mesh()
    for entity in msp:
        if entity.dxftype() == "LINE":
            start = entity.dxf.start
            end = entity.dxf.end
            vertices1 = [(start[0], start[1],  0), (end[0], end[1],  0)]
            vertices2 = [(start[0], start[1],  thickness), (end[0], end[1],  thickness)]
            
            face_vertices = vertices1 + vertices2
            with mesh.edit_data() as mesh_data:
                mesh_data.add_face(face_vertices)
                mesh_data.optimize()

    new_doc.saveas(output_file)

**Comment by mozman** [comment]
Hi!

`Ezdxf` is not the right tool for these tasks. I tried this myself and it involved a lot of rework in BricsCAD.

My recommendation are packages designed for these tasks:
- [build123d](https://github.com/gumyr/build123d)
- [solidpython2](https://pypi.org/project/solidpython2/)

**Reply by manpowre** [reply]
Thank you for these tips. yeah, I think I came to the end with ezdzf, atleast for the 3d part. For the dxf part to read the files I might still use ezdxf. again thanks

**Comment by manpowre** [comment]
<img width="1048" alt="Screenshot 2024-02-25 192500" src="https://github.com/mozman/ezdxf/assets/29288793/1c6dfe63-ff9a-40cf-b38c-482703f2e7ec">

I messed around a bit more with ARC to faces, but had to split the ARC into 10 segments with some math, and it's not exactly what I was looking for, so I figured for now ezDXF is not the right tool.
